File: packages/inputlayer-py/src/inputlayer/aggregations.py
"""Aggregation functions that compile to IQL aggregates."""

from __future__ import annotations

from typing import TYPE_CHECKING

from inputlayer._ast import AggExpr, Expr
from inputlayer._proxy import ColumnProxy

if TYPE_CHECKING:
    from inputlayer.relation import Relation


def _to_expr(col: ColumnProxy | Expr) -> Expr:
    if isinstance(col, ColumnProxy):
        return col._to_ast()
    return col


def count(column: ColumnProxy | type[Relation] | None = None) -> AggExpr:
    """Count rows. If a column is given, counts non-null values.

    IQL: count<Var>
    """
    if column is None or (isinstance(column, type)):
        # count(*) - needs at least one column from the body
        return AggExpr(func="count", column=None)
    return AggExpr(func="count", column=_to_expr(column))


def count_distinct(column: ColumnProxy) -> AggExpr:
    """Count distinct values.

    IQL: count_distinct<Var>
    """
    return AggExpr(func="count_distinct", column=_to_expr(column))


def sum_(column: ColumnProxy) -> AggExpr:
    """Sum numeric values.

    IQL: sum<Var>
    """
    return AggExpr(func="sum", column=_to_expr(column))


def min_(column: ColumnProxy) -> AggExpr:
    """Minimum value.

    IQL: min<Var>
    """
    return AggExpr(func="min", column=_to_expr(column))


def max_(column: ColumnProxy) -> AggExpr:
    """Maximum value.

    IQL: max<Var>
    """
    return AggExpr(func="max", column=_to_expr(column))


def avg(column: ColumnProxy) -> AggExpr:
    """Average value.

    IQL: avg<Var>
    """
    return AggExpr(func="avg", column=_to_expr(column))


def top_k(
    k: int,
    *passthrough: ColumnProxy,
    order_by: ColumnProxy,
    desc: bool = True,
) -> AggExpr:
    """Top-K aggregation with ordering.

    IQL: top_k<k, Passthrough..., OrderCol:desc>
    """
    return AggExpr(
        func="top_k",
        params=(k,),
        passthrough=tuple(_to_expr(p) for p in passthrough),
        order_column=_to_expr(order_by),
        desc=desc,
    )


def top_k_threshold(
    k: int,
    threshold: float,
    *passthrough: ColumnProxy,
    order_by: ColumnProxy,
    desc: bool = True,
) -> AggExpr:
    """Top-K with threshold aggregation.

    IQL: top_k_threshold<k, threshold, Passthrough..., OrderCol:desc>
    """
    return AggExpr(
        func="top_k_threshold",
        params=(k, threshold),
        passthrough=tuple(_to_expr(p) for p in passthrough),
        order_column=_to_expr(order_by),
        desc=desc,
    )


def within_radius(
    max_distance: float,
    *passthrough: ColumnProxy,
    distance: ColumnProxy | Expr,
    asc: bool = True,
) -> AggExpr:
    """Within-radius aggregation.

    IQL: within_radius<r, Passthrough..., DistCol:asc>
    """
    return AggExpr(
        func="within_radius",
        params=(max_distance,),
        passthrough=tuple(_to_expr(p) for p in passthrough),
        order_column=_to_expr(distance),
        desc=not asc,
    )

File: packages/inputlayer-js/src/aggregations.ts
/**
 * Aggregation functions that compile to IQL aggregates.
 */

import { type Expr, type AggExpr, aggExpr } from './ast.js';
import type { ColumnProxy } from './proxy.js';
import { wrap } from './proxy.js';

function toExpr(col: ColumnProxy | Expr): Expr {
  if ('toAst' in col && typeof (col as ColumnProxy).toAst === 'function') {
    return (col as ColumnProxy).toAst();
  }
  return col as Expr;
}

/**
 * Count rows. If a column is given, counts non-null values.
 * IQL: count<Var>
 */
export function count(column?: ColumnProxy | Expr): AggExpr {
  if (column === undefined) {
    return aggExpr({ func: 'count' });
  }
  return aggExpr({ func: 'count', column: toExpr(column) });
}

/**
 * Count distinct values.
 * IQL: count_distinct<Var>
 */
export function countDistinct(column: ColumnProxy | Expr): AggExpr {
  return aggExpr({ func: 'count_distinct', column: toExpr(column) });
}

/**
 * Sum numeric values.
 * IQL: sum<Var>
 */
export function sum(column: ColumnProxy | Expr): AggExpr {
  return aggExpr({ func: 'sum', column: toExpr(column) });
}

/**
 * Minimum value.
 * IQL: min<Var>
 */
export function min(column: ColumnProxy | Expr): AggExpr {
  return aggExpr({ func: 'min', column: toExpr(column) });
}

/**
 * Maximum value.
 * IQL: max<Var>
 */
export function max(column: ColumnProxy | Expr): AggExpr {
  return aggExpr({ func: 'max', column: toExpr(column) });
}

/**
 * Average value.
 * IQL: avg<Var>
 */
export function avg(column: ColumnProxy | Expr): AggExpr {
  return aggExpr({ func: 'avg', column: toExpr(column) });
}

/**
 * Top-K aggregation with ordering.
 * IQL: top_k<k, Passthrough..., OrderCol:desc>
 */
export function topK(opts: {
  k: number;
  passthrough?: Array<ColumnProxy | Expr>;
  orderBy: ColumnProxy | Expr;
  desc?: boolean;
}): AggExpr {
  return aggExpr({
    func: 'top_k',
    params: [opts.k],
    passthrough: (opts.passthrough ?? []).map(toExpr),
    orderColumn: toExpr(opts.orderBy),
    desc: opts.desc ?? true,
  });
}

/**
 * Top-K with threshold aggregation.
 * IQL: top_k_threshold<k, threshold, Passthrough..., OrderCol:desc>
 */
export function topKThreshold(opts: {
  k: number;
  threshold: number;
  passthrough?: Array<ColumnProxy | Expr>;
  orderBy: ColumnProxy | Expr;
  desc?: boolean;
}): AggExpr {
  return aggExpr({
    func: 'top_k_threshold',
    params: [opts.k, opts.threshold],
    passthrough: (opts.passthrough ?? []).map(toExpr),
    orderColumn: toExpr(opts.orderBy),
    desc: opts.desc ?? true,
  });
}

/**
 * Within-radius aggregation.
 * IQL: within_radius<r, Passthrough..., DistCol:asc>
 */
export function withinRadius(opts: {
  maxDistance: number;
  passthrough?: Array<ColumnProxy | Expr>;
  distance: ColumnProxy | Expr;
  asc?: boolean;
}): AggExpr {
  return aggExpr({
    func: 'within_radius',
    params: [opts.maxDistance],
    passthrough: (opts.passthrough ?? []).map(toExpr),
    orderColumn: toExpr(opts.distance),
    desc: !(opts.asc ?? true),
  });
}
